src/eventbridge/eventpattern/index.ts

```diff
@@ -0,0 +1,798 @@
+    right: Expr,
+    op: BinaryOp
+  ): {
+    eventReference: EventReference;
+    eventExpr: PropAccessExpr | ElementAccessExpr;
+    other: Expr;
+    op: BinaryOp;
+  } => {
+    const leftExpr = getEventReference(left);
+    const rightExpr = getEventReference(right);
+    if (Array.isArray(leftExpr) && Array.isArray(rightExpr)) {
+      throw new Error("Expected exactly one event reference, got two.");
+    } else if (Array.isArray(leftExpr)) {
+      assertValidEventRefererence(leftExpr);
+      return {
+        eventReference: leftExpr,
+        eventExpr: left as PropAccessExpr | ElementAccessExpr,
+        other: right,
+        op,
+      };
+    } else if (Array.isArray(rightExpr)) {
+      assertValidEventRefererence(rightExpr);
+      return {
+        eventReference: rightExpr,
+        eventExpr: right as PropAccessExpr | ElementAccessExpr,
+        other: left,
+        op: invertBinaryOperator(op),
+      };
+    }
+    throw new Error("Expected exactly one event reference, got zero.");
+  };
+
+  /**
+   * Recurse an expression to find a reference to the event.
+   */
+  const getEventReference = (expression: Expr): string[] | undefined => {
+    if (isIdentifier(expression)) {
+      if (expression.name === eventDecl.name) {
+        return [];
+      }
+      // TODO: support references
+      throw Error(
+        `All identifiers must point to the event parameter, found: ${expression.name}`
+      );
+    } else if (
+      isPropAccessExpr(expression) ||
+      isElementAccessExpr(expression)
+    ) {
+      const key = getPropertyAccessKey(expression);
+      const parent = getEventReference(expression.expr);
+      if (parent) {
+        return [...parent, key];
+      }
+      return undefined;
+    }
+    return undefined;
+  };
+
+  // find the return, we'll the resolve the rest as needed.
+  const [ret] = predicate.body.statements.filter((expr) => isReturn(expr));
+
+  if (!ret) {
+    throw new Error("Missing return statement in predicate block.");
+  }
+
+  const result = handleSmnt(ret);
+
+  return patternDocumentToEventPattern(result);
+};
+
+type EventReference = [string] | ["detail", ...string[]];
+
+// TODO: validate again object schema?
+function assertValidEventRefererence(
+  eventReference: string[]
+): asserts eventReference is EventReference {
+  if (eventReference.length === 0) {
+    throw new Error("Direct use of the event is invalid.");
+  }
+  if (eventReference.length > 1) {
+    const [first] = eventReference;
+    if (first !== "detail") {
+      throw `Event references with depth greater than one must be on the detail propert, got ${eventReference.join(
+        ","
+      )}`;
+    }
+  }
+}
+
+/**
+ * Retrieves a string, number, boolean, undefined, or null constant from the given expression.
+ * Wrap the value to not be ambiguous with the undefined value.
+ * When one is not found, return undefined (not wrapped).
+ *
+ * TODO: Support following references.
+ */
+const getConstant = (
```

Comment:
this looks like an `eval` function - it evals the `Expr` to a literal value? Call it `evalToConstant`?